### ISSUE TITLE
Bad Design (or what) - the sleeping thread reuse phenomenon

### ISSUE THREAD

**Issue by KrishnaPG** [Microsoft/cpprestsdk#174]
Trying out this PPLX from this CPPRest SDK to validate its scalability, realize that it is creating new threads (as many threads as the number of pending requests?), instead of "scheduling" them onto a pool.

Created a simple "sleeping" thread long-operation and called it from JS 20 times in a row - and ended up with 20+ threads (all sleeping), where in I would have expected the requests to be queued to a fixed set of pool (as many as the no. of cores), and sleeping threads just to switch the context and take care of new requests (something like IOCP)
````
void longop(int a)
{
	pplx::create_task([=]() {
			boost::this_thread::sleep(boost::posix_time::milliseconds(a * 100000));	
		return a;
	}).then([=](uint64_t i) {
		sendResult(std::make_tuple(i));
	});
}
````

Here is the visuals: all 20+ threads sleeping happily
![image](https://cloud.githubusercontent.com/assets/4610831/15919269/3d493da8-2e2d-11e6-8d37-c50b9b4d88d4.png)

In other words, if 10,000 clients connect - we see 10K threads?? A very pre c10K era design

**Comment by ras0219-msft** [comment]
On Windows, we schedule tasks onto the OS threadpool. Could you retry your experiment with a more "real" load (something that actually consumes CPU)? It is very possible that the Windows Threadpool is "outsmarting" the benchmark here.

**Comment by KrishnaPG** [comment]
Interesting. Replaced the `sleep` with some `while` loop and the thread count did not go beyond 5 or 6, and the pending requests are waiting (not getting scheduled).

So, what determines this "new thread" creation? How is it that while sleeping, new threads are created and while doing some CPU intensive job no further new threads are created (and rather incoming requests are kept waiting)?

**Comment by ras0219-msft** [comment]
The Windows Threadpool is pretty intelligent and will adapt to the workload. In your sleeping example, it determined that massive oversubscription was the fastest way to accomplish your workload. In the CPU-bound case, it knows oversubscription would reduce performance and so caps the number of threads at an appropriate level 😄.

In I/O cases (modulo bugs in the C++ REST SDK), it should result in the high performance behavior you've described above: very few threads servicing a ton of IOCP requests.
